File: clearcore/src/command_handler.cpp
/**
 * Space Maquette - Command Handler Implementation
 */

#include "command_handler.h"

CommandHandler::CommandHandler(CommandParser& parser, MotionControl& motion,
                               Rangefinder& rangefinder, EmergencyStop& estop,
                               ConfigurationManager& config)
    : _parser(parser),
      _motion(motion),
      _rangefinder(rangefinder),
      _estop(estop),
      _config(config),
      _debugMode(false) {}

void CommandHandler::init() {
    // Register this handler with the parser
    _parser.setCommandHandler(
        [this](CommandParser& parser) -> void { this->processCommand(parser); });

#ifdef DEBUG
    Serial.println("Command handler initialized");
#endif
}

void CommandHandler::processCommand(CommandParser& parser) {
    const char* cmd = _parser.getCommand();
    if (!cmd)
        return;

#ifdef DEBUG
    Serial.print("Processing command: ");
    Serial.println(cmd);
#endif

    // Check for ESTOP first
    if (strcmp(cmd, "ESTOP") == 0) {
        _estop.activate();
        _parser.sendResponse("OK", "ESTOP_ACTIVATED");
        return;
    }

    // If ESTOP is active, only allow certain commands
    if (_estop.isActive() && strcmp(cmd, "STATUS") != 0 && strcmp(cmd, "RESET_ESTOP") != 0) {
        _parser.sendResponse("ERROR", "ESTOP_ACTIVE");
        return;
    }

    // Handle RESET_ESTOP
    if (strcmp(cmd, "RESET_ESTOP") == 0) {
        bool success = _estop.reset();
        if (success) {
            _parser.sendResponse("OK", "ESTOP_RESET");
        } else {
            _parser.sendResponse("ERROR", "ESTOP_STILL_ACTIVE");
        }
        return;
    }

    // Process command by category
    if (strcmp(cmd, "PING") == 0) {
        _parser.sendResponse("OK", "PONG");
    } else if (strcmp(cmd, "RESET") == 0 || strcmp(cmd, "STATUS") == 0 ||
               strcmp(cmd, "DEBUG") == 0) {
        handleSystemCommands();
    } else if (strcmp(cmd, "HOME") == 0 || strcmp(cmd, "MOVE") == 0 || strcmp(cmd, "STOP") == 0 ||
               strcmp(cmd, "VELOCITY") == 0) {
        handleMotionCommands();
    } else if (strcmp(cmd, "MEASURE") == 0 || strcmp(cmd, "SCAN") == 0) {
        handleRangefinderCommands();
    } else if (strcmp(cmd, "TILT") == 0 || strcmp(cmd, "PAN") == 0) {
        handleServoCommands();
    } else if (strcmp(cmd, "CONFIG") == 0 || strcmp(cmd, "GET") == 0 || strcmp(cmd, "SET") == 0 ||
               strcmp(cmd, "SAVE") == 0) {
        handleConfigCommands();
    } else {
        _parser.sendResponse("ERROR", "UNKNOWN_COMMAND");
    }
}
void CommandHandler::handleSystemCommands() {
    const char* cmd = _parser.getCommand();

    if (strcmp(cmd, "PING") == 0) {
        _parser.sendResponse("OK", "PONG");
    } else if (strcmp(cmd, "RESET") == 0) {
        // Soft reset functionality
        _parser.sendResponse("OK", "RESETTING");

        // Reset subsystems
        _motion.stop();

#ifdef DEBUG
        Serial.println("System reset");
#endif
    } else if (strcmp(cmd, "STATUS") == 0) {
        // Get current system status
        char statusBuffer[128];
        sprintf(statusBuffer, "X=%.2f,Y=%.2f,Z=%.2f,PAN=%.2f,TILT=%.2f,ESTOP=%d,MOVING=%d,HOMED=%d",
                _motion.getPositionX(), _motion.getPositionY(), _motion.getPositionZ(),
                _motion.getPanAngle(), _motion.getTiltAngle(), _estop.isActive() ? 1 : 0,
                _motion.isMoving() ? 1 : 0, _motion.isHomed() ? 1 : 0);

        _parser.sendResponse("OK", statusBuffer);
    } else if (strcmp(cmd, "DEBUG") == 0) {
        if (_parser.getParamCount() > 0) {
            const char* mode = _parser.getParam(0);
            if (strcmp(mode, "ON") == 0) {
                _debugMode = true;
                _rangefinder.setDebug(
                    true);  // Assuming this method exists based on _debugEnabled in rangefinder.cpp
                _parser.sendResponse("OK", "DEBUG_ENABLED");
            } else if (strcmp(mode, "OFF") == 0) {
                _debugMode = false;
                _parser.sendResponse("OK", "DEBUG_DISABLED");
            } else {
                _parser.sendResponse("ERROR", "INVALID_PARAM");
            }
        } else {
            _parser.sendResponse("ERROR", "MISSING_PARAM");
        }
    }
}

void CommandHandler::handleMotionCommands() {
    const char* cmd = _parser.getCommand();

    if (strcmp(cmd, "HOME") == 0) {
        if (_parser.getParamCount() > 0) {
            const char* axis = _parser.getParam(0);
            bool success = false;

            if (strcmp(axis, "ALL") == 0) {
                success = _motion.homeAllAxes();
            } else if (strcmp(axis, "X") == 0) {
                success = _motion.homeAxis('X');
            } else if (strcmp(axis, "Y") == 0) {
                success = _motion.homeAxis('Y');
            } else if (strcmp(axis, "Z") == 0) {
                success = _motion.homeAxis('Z');
            } else {
                _parser.sendResponse("ERROR", "INVALID_AXIS");
                return;
            }

            if (success) {
                _parser.sendResponse("OK", "HOMING_STARTED");
            } else {
                _parser.sendResponse("ERROR", "HOMING_FAILED");
            }
        } else {
            _parser.sendResponse("ERROR", "MISSING_PARAM");
        }
    } else if (strcmp(cmd, "MOVE") == 0) {
        if (_parser.getParamCount() >= 3) {
            float x = _parser.getParamAsFloat(0);
            float y = _parser.getParamAsFloat(1);
            float z = _parser.getParamAsFloat(2);
            float pan =
                _parser.getParamCount() > 3 ? _parser.getParamAsFloat(3) : _motion.getPanAngle();
            float tilt =
                _parser.getParamCount() > 4 ? _parser.getParamAsFloat(4) : _motion.getTiltAngle();

            bool success = _motion.moveToPosition(x, y, z, pan, tilt);

            if (success) {
                _parser.sendResponse("OK", "MOVE_STARTED");
            } else {
                _parser.sendResponse("ERROR", "MOVE_FAILED");
            }
        } else {
            _parser.sendResponse("ERROR", "MISSING_PARAMS");
        }
    } else if (strcmp(cmd, "STOP") == 0) {
        _motion.stop();
        _parser.sendResponse("OK", "MOTION_STOPPED");
    } else if (strcmp(cmd, "VELOCITY") == 0) {
        if (_parser.getParamCount() >= 3) {
            float vx = _parser.getParamAsFloat(0);
            float vy = _parser.getParamAsFloat(1);
            float vz = _parser.getParamAsFloat(2);

            _motion.setVelocity(vx, vy, vz);
            _parser.sendResponse("OK", "VELOCITY_SET");
        } else {
            _parser.sendResponse("ERROR", "MISSING_PARAMS");
        }
    }
}

void CommandHandler::handleRangefinderCommands() {
    const char* cmd = _parser.getCommand();

    if (strcmp(cmd, "MEASURE") == 0) {
        float distance = _rangefinder.takeMeasurement();

        if (distance >= 0) {
            _parser.sendFormattedResponse("OK", "%.3f", distance);
        } else if (distance == -2.0) {
            _parser.sendResponse("ERROR", "OUT_OF_RANGE");
        } else {
            _parser.sendResponse("ERROR", "MEASUREMENT_FAILED");
        }
    } else if (strcmp(cmd, "SCAN") == 0) {
        if (_parser.getParamCount() >= 5) {
#ifdef DEBUG
            float x1 = _parser.getParamAsFloat(0);
            float y1 = _parser.getParamAsFloat(1);
            float x2 = _parser.getParamAsFloat(2);
            float y2 = _parser.getParamAsFloat(3);
            float step = _parser.getParamAsFloat(4);
#endif

            // Start scan operation
            // This would typically be implemented as a state machine
            // that executes over multiple loop iterations

            _parser.sendResponse("OK", "SCAN_STARTED");

#ifdef DEBUG
            Serial.print("Scan requested: (");
            Serial.print(x1);
            Serial.print(",");
            Serial.print(y1);
            Serial.print(") to (");
            Serial.print(x2);
            Serial.print(",");
            Serial.print(y2);
            Serial.print(") with step ");
            Serial.println(step);
#endif
        } else {
            _parser.sendResponse("ERROR", "MISSING_PARAMS");
        }
    }
}

void CommandHandler::handleServoCommands() {
    const char* cmd = _parser.getCommand();

    if (strcmp(cmd, "TILT") == 0) {
        if (_parser.getParamCount() > 0) {
            float angle = _parser.getParamAsFloat(0);
            bool success = _motion.setPanAngle(static_cast<int32_t>(angle));

            if (success) {
                _parser.sendResponse("OK", "TILT_SET");
            } else {
                _parser.sendResponse("ERROR", "TILT_FAILED");
            }
        } else {
            _parser.sendResponse("ERROR", "MISSING_PARAM");
        }
    } else if (strcmp(cmd, "PAN") == 0) {
        if (_parser.getParamCount() > 0) {
            float angle = _parser.getParamAsFloat(0);
            bool success = _motion.setPanAngle(
                static_cast<int32_t>(angle));  // angle is float, but method expects int32_t

            if (success) {
                _parser.sendResponse("OK", "PAN_SET");
            } else {
                _parser.sendResponse("ERROR", "PAN_FAILED");
            }
        } else {
            _parser.sendResponse("ERROR", "MISSING_PARAM");
        }
    }
}

void CommandHandler::handleConfigCommands() {
    const char* cmd = _parser.getCommand();

    if (strcmp(cmd, "CONFIG") == 0) {
        if (_parser.getParamCount() > 0) {
            const char* subCmd = _parser.getParam(0);

            if (strcmp(subCmd, "LOAD") == 0) {
                bool success = _config.loadConfig();
                if (success) {
                    _parser.sendResponse("OK", "CONFIG_LOADED");
                } else {
                    _parser.sendResponse("ERROR", "CONFIG_LOAD_FAILED");
                }
            } else if (strcmp(subCmd, "SAVE") == 0) {
                bool success = _config.saveConfig();
                if (success) {
                    _parser.sendResponse("OK", "CONFIG_SAVED");
                } else {
                    _parser.sendResponse("ERROR", "CONFIG_SAVE_FAILED");
                }
            } else if (strcmp(subCmd, "LIST") == 0) {
                // This would require adding a method to list all configuration items
                // For now, just acknowledge the command
                _parser.sendResponse("OK", "CONFIG_LIST_NOT_IMPLEMENTED");
            } else {
                _parser.sendResponse("ERROR", "INVALID_CONFIG_COMMAND");
            }
        } else {
            _parser.sendResponse("ERROR", "MISSING_CONFIG_COMMAND");
        }
    } else if (strcmp(cmd, "GET") == 0) {
        if (_parser.getParamCount() > 0) {
            const char* key = _parser.getParam(0);

            if (_config.hasKey(key)) {
                String value = _config.getString(key, "");
                _parser.sendResponse("OK", value.c_str());
            } else {
                _parser.sendResponse("ERROR", "KEY_NOT_FOUND");
            }
        } else {
            _parser.sendResponse("ERROR", "MISSING_KEY");
        }
    } else if (strcmp(cmd, "SET") == 0) {
        if (_parser.getParamCount() >= 2) {
            const char* key = _parser.getParam(0);
            const char* value = _parser.getParam(1);

            _config.setString(key, value);
            _parser.sendResponse("OK", "VALUE_SET");

            // Apply certain configuration values immediately
            if (strcmp(key, "tilt_min") == 0) {
                _motion.setTiltLimits(_config.getInt("tilt_min", 45),
                                      _config.getInt("tilt_max", 135));
            } else if (strcmp(key, "tilt_max") == 0) {
                _motion.setTiltLimits(_config.getInt("tilt_min", 45),
                                      _config.getInt("tilt_max", 135));
            } else if (strcmp(key, "velocity_x") == 0) {
                _motion.setVelocity(_config.getInt("velocity_x", DEFAULT_VELOCITY_LIMIT),
                                    _config.getInt("velocity_y", DEFAULT_VELOCITY_LIMIT),
                                    _config.getInt("velocity_z", DEFAULT_VELOCITY_LIMIT));
            }
            // Add more immediate application cases as needed

        } else {
            _parser.sendResponse("ERROR", "MISSING_PARAMS");
        }
    } else if (strcmp(cmd, "SAVE") == 0) {
        bool success = _config.saveConfig();
        if (success) {
            _parser.sendResponse("OK", "CONFIG_SAVED");
        } else {
            _parser.sendResponse("ERROR", "CONFIG_SAVE_FAILED");
        }
    }
}

File: clearcore_comms_test/clearcore_comms_test/client/command_handler.py
"""
Command handler for the ClearCore controller.
Provides high-level methods for sending commands to the ClearCore.
"""
from typing import List, Optional, Tuple, Union, Dict, Any
from .ethernet_client import ClearCoreClient

class ClearCoreCommandHandler:
    """High-level command handler for the ClearCore controller."""
    
    def __init__(self, client: ClearCoreClient):
        """
        Initialize the command handler.
        
        Args:
            client: ClearCoreClient instance
        """
        self.client = client
    
    # System Commands
    
    def ping(self) -> Tuple[bool, str]:
        """
        Send a ping command to check if the system is responsive.
        
        Returns:
            Tuple of (success, response)
        """
        return self.client.send_command("PING")
    
    def reset(self) -> Tuple[bool, str]:
        """
        Perform a soft reset of the system.
        
        Returns:
            Tuple of (success, response)
        """
        return self.client.send_command("RESET")
    
    def get_status(self) -> Tuple[bool, Dict[str, Any]]:
        """
        Get current system status.
        
        Returns:
            Tuple of (success, status_dict)
        """
        success, response = self.client.send_command("STATUS")
        if success:
            # Parse status response
            # Format: OK:X=<x>,Y=<y>,Z=<z>,PAN=<pan>,TILT=<tilt>,ESTOP=<0/1>,MOVING=<0/1>,HOMED=<0/1>
            try:
                _, status_str = response.split(":", 1)
                status_parts = status_str.split(",")
                status_dict = {}
                
                for part in status_parts:
                    key, value = part.split("=", 1)
                    # Convert numeric values
                    if key in ["X", "Y", "Z", "PAN", "TILT"]:
                        status_dict[key.lower()] = float(value)
                    elif key in ["ESTOP", "MOVING", "HOMED"]:
                        status_dict[key.lower()] = value == "1"
                    else:
                        status_dict[key.lower()] = value
                
                return True, status_dict
            except (ValueError, IndexError):
                return False, {"error": "Failed to parse status response"}
        else:
            return False, {"error": response}
    
    def set_debug(self, enabled: bool) -> Tuple[bool, str]:
        """
        Enable or disable debug mode.
        
        Args:
            enabled: True to enable debug mode, False to disable
            
        Returns:
            Tuple of (success, response)
        """
        mode = "ON" if enabled else "OFF"
        return self.client.send_command("DEBUG", [mode])
    
    def activate_estop(self) -> Tuple[bool, str]:
        """
        Activate emergency stop.
        
        Returns:
            Tuple of (success, response)
        """
        return self.client.send_command("ESTOP")
    
    def reset_estop(self) -> Tuple[bool, str]:
        """
        Reset emergency stop if safe.
        
        Returns:
            Tuple of (success, response)
        """
        return self.client.send_command("RESET_ESTOP")
    
    # Motion Commands
    
    def home(self, axis: str = "ALL") -> Tuple[bool, str]:
        """
        Home specified axis or all axes.
        
        Args:
            axis: Axis to home ("ALL", "X", "Y", or "Z")
            
        Returns:
            Tuple of (success, response)
        """
        if axis not in ["ALL", "X", "Y", "Z"]:
            return False, "ERROR:INVALID_AXIS"
        
        return self.client.send_command("HOME", [axis])
    
    def move(self, x: float, y: float, z: float, 
             pan: Optional[float] = None, tilt: Optional[float] = None) -> Tuple[bool, str]:
        """
        Move to absolute position.
        
        Args:
            x: X position
            y: Y position
            z: Z position
            pan: Pan angle (optional)
            tilt: Tilt angle (optional)
            
        Returns:
            Tuple of (success, response)
        """
        params = [x, y, z]
        if pan is not None:
            params.append(pan)
            if tilt is not None:
                params.append(tilt)
        
        return self.client.send_command("MOVE", params)
    
    def stop(self) -> Tuple[bool, str]:
        """
        Stop all motion immediately.
        
        Returns:
            Tuple of (success, response)
        """
        return self.client.send_command("STOP")
    
    def set_velocity(self, vx: float, vy: float, vz: float) -> Tuple[bool, str]:
        """
        Set axis velocities.
        
        Args:
            vx: X axis velocity
            vy: Y axis velocity
            vz: Z axis velocity
            
        Returns:
            Tuple of (success, response)
        """
        return self.client.send_command("VELOCITY", [vx, vy, vz])
    
    # Rangefinder Commands
    
    def measure(self) -> Tuple[bool, Union[float, str]]:
        """
        Take a single distance measurement.
        
        Returns:
            Tuple of (success, distance or error message)
        """
        success, response = self.client.send_command("MEASURE")
        if success:
            try:
                _, distance_str = response.split(":", 1)
                return True, float(distance_str)
            except (ValueError, IndexError):
                return False, "Failed to parse measurement response"
        else:
            return False, response
    
    def scan(self, x1: float, y1: float, x2: float, y2: float, step: float) -> Tuple[bool, str]:
        """
        Perform a scan over the specified area.
        
        Args:
            x1: Start X position
            y1: Start Y position
            x2: End X position
            y2: End Y position
            step: Step size
            
        Returns:
            Tuple of (success, response)
        """
        return self.client.send_command("SCAN", [x1, y1, x2, y2, step])
    
    # Servo Commands
    
    def set_tilt(self, angle: float) -> Tuple[bool, str]:
        """
        Set tilt servo angle.
        
        Args:
            angle: Tilt angle
            
        Returns:
            Tuple of (success, response)
        """
        return self.client.send_command("TILT", [angle])
    
    def set_pan(self, angle: float) -> Tuple[bool, str]:
        """
        Set pan axis angle.
        
        Args:
            angle: Pan angle
            
        Returns:
            Tuple of (success, response)
        """
        return self.client.send_command("PAN", [angle])
    
    # Configuration Commands
    
    def load_config(self) -> Tuple[bool, str]:
        """
        Load configuration from SD card.
        
        Returns:
            Tuple of (success, response)
        """
        return self.client.send_command("CONFIG", ["LOAD"])
    
    def save_config(self) -> Tuple[bool, str]:
        """
        Save configuration to SD card.
        
        Returns:
            Tuple of (success, response)
        """
        return self.client.send_command("CONFIG", ["SAVE"])
    
    def list_config(self) -> Tuple[bool, str]:
        """
        List all configuration items.
        
        Returns:
            Tuple of (success, response)
        """
        return self.client.send_command("CONFIG", ["LIST"])
    
    def get_config(self, key: str) -> Tuple[bool, str]:
        """
        Get configuration value.
        
        Args:
            key: Configuration key
            
        Returns:
            Tuple of (success, value or error message)
        """
        return self.client.send_command("GET", [key])
    
    def set_config(self, key: str, value: str) -> Tuple[bool, str]:
        """
        Set configuration value.
        
        Args:
            key: Configuration key
            value: Configuration value
            
        Returns:
            Tuple of (success, response)
        """
        return self.client.send_command("SET", [key, value])
